packages/dds/sequence/src/test/endpointInRangeIndex.spec.ts

```diff
@@ -0,0 +1,186 @@
+	private readonly intervals: Interval[];
+
+	constructor() {
+		this.intervals = [];
+	}
+
+	add(interval: Interval) {
+		this.intervals.push(interval);
+	}
+
+	findIntervalsWithEndpointInRange(start: number, end: number): Interval[] {
+		return this.intervals.filter((interval) => interval.end >= start && interval.end <= end);
+	}
+}
+
+function createTestInterval(start: number, end: number): Interval {
+	const props: PropertySet = {};
+	props[reservedIntervalIdKey] = [uuid()];
+
+	return new Interval(start, end, props);
+}
+
+function assertPlainNumberIntervals(
+	results: Interval[],
+	expectedEndpoints: { start: number; end: number }[] | Interval[],
+): void {
+	assert.equal(results.length, expectedEndpoints.length, "Mismatched result count");
+	for (let i = 0; i < results.length; ++i) {
+		assert(results[i]);
+		assert.equal(results[i].start, expectedEndpoints[i].start, "mismatched start");
+		assert.equal(results[i].end, expectedEndpoints[i].end, "mismatched end");
+	}
+}
+
+function generateRandomIntervals(count: number, rangeStart: number, rangeEnd: number) {
+	const intervals: Interval[] = [];
+
+	while (intervals.length < count) {
+		const start = getRandomNumber(rangeStart, rangeEnd);
+		const end = getRandomNumber(start, rangeEnd);
+		const interval = createTestInterval(start, end);
+		/**
+		 * Currently avoid using duplicate intervals due to the bug:
+		 * https://dev.azure.com/fluidframework/internal/_workitems/edit/4477
+		 */
+		if (
+			!intervals.some(
+				(existing) => existing.start === interval.start && existing.end === interval.end,
+			)
+		) {
+			intervals.push(interval);
+		}
+	}
+
+	return intervals;
+}
+
+function getRandomNumber(min: number, max: number): number {
+	return Math.floor(Math.random() * (max - min + 1)) + min;
+}
+
+describe("Support querying intervals with endpoints in a specified range", () => {
+	let helpers: IIntervalHelpers<Interval>;
+	let endpointInRangeIndex;
+	let compareFn;
+	let results;
+
+	beforeEach(() => {
+		helpers = {
+			compareEnds: (a: Interval, b: Interval) => a.end - b.end,
+			create: createInterval,
+		};
+		endpointInRangeIndex = createEndpointInRangeIndex(helpers, undefined as any as Client);
+		// sort the query result by the interval endpoint value
+		compareFn = (a: Interval, b: Interval) => {
+			if (a.end === b.end) {
+				return a.start - b.start;
+			}
+			return a.end - b.end;
+		};
+	});
+
+	it("return empty result when the index is empty", () => {
+		results = endpointInRangeIndex.findIntervalsWithEndpointInRange(1, 1);
+		assert.equal(results.length, 0, "Should not return anything once the index is empty");
+	});
+
+	it("limit on the start/end value", () => {
+		endpointInRangeIndex.add(createTestInterval(1, 1));
+		endpointInRangeIndex.add(createTestInterval(1, 3));
+
+		results = endpointInRangeIndex.findIntervalsWithEndpointInRange(2, 1);
+		assert.equal(results.length, 0, "The start value should not be lower than the end value");
```

Comment:
nit: I don't think the assertion message really makes much sense here. Assertion messages are displayed to the test runner when the corresponding test fails. Here you're really checking that finding overlapping intervals with start > end should never return results.

I think part of the reason the error message is less clear is that you're trying to test multiple things per test. If you try to structure your tests such that:

1. Each test tests one thing
2. The test's fully qualified name (i.e. including suite blocks above it) gives a description of what the test is testing

you can lean a lot less heavily on having string assertions. For example, one way you could structure this test as 2 tests and drop all of the assert messages would be:

```typescript
describe("findIntervalsWithEndpointInRange", () => {
    describe("finds no intervals", () => {
        it("when start > end for the query range");
        it("when all intervals are above the query range");
    });
});
```
I only wrote those 2 tests because I'm not sure how much value there is in testing negative behavior. It might be interesting, but in theory it shouldn't be (it's totally a valid thing to have negative intervals). But you could name that test case `"when all intervals are above a negative query range"`.

This test block would also naturally contain the test you wrote above: `"when the index is empty"`.

When you make this kind of refactor, it also helps spot test cases you might have missed. For example: `"when all intervals are below the query range"`, and `"when the query range is between two intervals' endpoints"`

Reply:
refactored the test blocks in the latest commit https://github.com/microsoft/FluidFramework/pull/15837/commits/ed12bbc0e4de5f1b69082588917026aa852c1521